### ISSUE TITLE
source/DWGUGens/dwglib/DWG.hpp fail to build with GCC 6

### ISSUE THREAD

**Issue by petterreinholdtsen** [supercollider/sc3-plugins#106]
The source fail to build in Debian after Debian switched to GCC 6 as the default C++ compiler.  See https://bugs.debian.org/831078 for details.

The problem is that this expression no longer work in source/DWGUGens/dwglib/DWG.hpp:
```
float loss;
...
void set_area(float a1,float lossF)
    {
        if(a1 < 1e-18)
            loss = 0.0;
        else
            loss = 1.0 - std::min(lossF/sqrt(static_cast<double>(a1)),1.0);

    }
```

I believe the cause is that the compiler now pick the function float sqrt(float) instead of the function double sqrt(double).  There are two ways to solve the build problem, and I am unsure which one is the best.  Either replace 1.0 with 1.0f to use float everywhere in the calculations, or cast a1 using static_cast(double>(a1) to tell the compiler to use doubles everywhere.  I assume the former calculation is slightly quicker, while the latter is more accurate.

Do you have an opinion on how this should be solved?  How important is the accuracy of the calculation?

**Comment by snappizz** [comment]
Paging @sonoro1234.

**Comment by sonoro1234** [comment]
I thing 1.0f is ok

**Comment by petterreinholdtsen** [comment]
Any hope to have a fix for this included in a new release?

**Comment by snappizz** [comment]
3.8 is in release mode, so #108 will likely be merged before then.

**Comment by hzulla** [comment]
@snappizz Sadly, this fix will break gcc 5. Possible patch (worked on my machine) is this:

```
#if defined(__GNUC__) && __GNUC__ >= 6
            loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
#else
            loss = 1.0 - std::min(lossF/sqrt(a1),1.0);
#endif
```

**Comment by sonoro1234** [comment]
@hzulla How did it break gcc5?

**Comment by hzulla** [comment]
```
In file included from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:8:0,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp: In member function 'void TUBE::set_area(float, float)':
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp:1271:54: error: no matching function for call to 'min(double, float)'
             loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
                                                      ^
In file included from /usr/include/c++/5/algorithm:61:0,
                 from /usr/include/SuperCollider/common/SC_BoundsMacros.h:26,
                 from /usr/include/SuperCollider/plugin_interface/SC_RGen.h:54,
                 from /usr/include/SuperCollider/plugin_interface/SC_World.h:28,
                 from /usr/include/SuperCollider/plugin_interface/SC_PlugIn.h:24,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:4,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/usr/include/c++/5/bits/stl_algobase.h:195:5: note: candidate: template<class _Tp> const _Tp& std::min(const _Tp&, const _Tp&)
     min(const _Tp& __a, const _Tp& __b)
     ^
/usr/include/c++/5/bits/stl_algobase.h:195:5: note:   template argument deduction/substitution failed:
In file included from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:8:0,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp:1271:54: note:   deduced conflicting types for parameter 'const _Tp' ('double' and 'float')
             loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
                                                      ^
In file included from /usr/include/c++/5/algorithm:61:0,
                 from /usr/include/SuperCollider/common/SC_BoundsMacros.h:26,
                 from /usr/include/SuperCollider/plugin_interface/SC_RGen.h:54,
                 from /usr/include/SuperCollider/plugin_interface/SC_World.h:28,
                 from /usr/include/SuperCollider/plugin_interface/SC_PlugIn.h:24,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:4,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/usr/include/c++/5/bits/stl_algobase.h:243:5: note: candidate: template<class _Tp, class _Compare> const _Tp& std::min(const _Tp&, const _Tp&, _Compare)
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^
/usr/include/c++/5/bits/stl_algobase.h:243:5: note:   template argument deduction/substitution failed:
In file included from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:8:0,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp:1271:54: note:   deduced conflicting types for parameter 'const _Tp' ('double' and 'float')
             loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
                                                      ^
In file included from /usr/include/c++/5/algorithm:62:0,
                 from /usr/include/SuperCollider/common/SC_BoundsMacros.h:26,
                 from /usr/include/SuperCollider/plugin_interface/SC_RGen.h:54,
                 from /usr/include/SuperCollider/plugin_interface/SC_World.h:28,
                 from /usr/include/SuperCollider/plugin_interface/SC_PlugIn.h:24,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:4,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/usr/include/c++/5/bits/stl_algo.h:3445:5: note: candidate: template<class _Tp> _Tp std::min(std::initializer_list<_Tp>)
     min(initializer_list<_Tp> __l)
     ^
/usr/include/c++/5/bits/stl_algo.h:3445:5: note:   template argument deduction/substitution failed:
In file included from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:8:0,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp:1271:54: note:   mismatched types 'std::initializer_list<_Tp>' and 'double'
             loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
                                                      ^
In file included from /usr/include/c++/5/algorithm:62:0,
                 from /usr/include/SuperCollider/common/SC_BoundsMacros.h:26,
                 from /usr/include/SuperCollider/plugin_interface/SC_RGen.h:54,
                 from /usr/include/SuperCollider/plugin_interface/SC_World.h:28,
                 from /usr/include/SuperCollider/plugin_interface/SC_PlugIn.h:24,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:4,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/usr/include/c++/5/bits/stl_algo.h:3451:5: note: candidate: template<class _Tp, class _Compare> _Tp std::min(std::initializer_list<_Tp>, _Compare)
     min(initializer_list<_Tp> __l, _Compare __comp)
     ^
/usr/include/c++/5/bits/stl_algo.h:3451:5: note:   template argument deduction/substitution failed:
In file included from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/dwgs.h:8:0,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/piano.h:8,
                 from /home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/OteyPiano.cpp:7:
/home/hanno/deb/supercollider-sc3-plugins/source/OteyPianoUGens/../DWGUGens/dwglib/DWG.hpp:1271:54: note:   mismatched types 'std::initializer_list<_Tp>' and 'double'
             loss = 1.0 - std::min(lossF/sqrt(a1),1.0f);
                                                      ^
source/CMakeFiles/OteyPianoUGens.dir/build.make:65: recipe for target 'source/CMakeFiles/OteyPianoUGens.dir/OteyPianoUGens/OteyPiano.cpp.o' failed
```

**Comment by sonoro1234** [comment]
It is strange because there are no doubles and sqrt from a float should be a float according to C++11
```
http://www.cplusplus.com/reference/cmath/sqrt/
```
could be we are not in c++11 ?

**Comment by sonoro1234** [comment]
Beasides, we did not have std::min but sc_min before!!!

**Comment by hzulla** [comment]
I'm going to upgrade the package soon, but the Debian source package is still on release 3.7.1, which still has std::min.

**Comment by sonoro1234** [comment]
PR #112 has sc_min so it will be ok

**Comment by philipchungtech** [comment]
> It is strange because there are no doubles and sqrt from a float should be a float according to C++11

Actually, `float std::sqrt(float)` is present in C++03. http://en.cppreference.com/w/cpp/numeric/math/sqrt

What I think is happening is that under GCC 5, including `math.h` does not include the C++ overloads of `sqrt`, only the C version of the function, which is `double sqrt(double)`. Under GCC 6, though, `math.h` includes the C++ versions into the global namespace. To my understanding this is *nonstandard*, so the correct thing to do would be either:

1. Use C's `float sqrtf(float)`
2. Change the include to `<cmath>` and use `std::sqrt` (note the namespace)

**Comment by sonoro1234** [comment]
That seems quite reasonable.
I cant test gcc5 neither gcc6 only 4.8.2 so I would stick to sc_min (already done in repo) that let compile althought we dont know if we are working with floats or doubles.

If any person feels insecure about not knowing if we have floats or doubles and can try gcc5 and gcc6 may be could post any of this solutions!!
